src/Groups/matrices/forms.jl

```diff
@@ -0,0 +1,461 @@
+   end
+end
+
+"""
+    corresponding_quadratic_form(Q::SesquilinearForm)
+Given a symmetric form `f`, returns the quadratic form `Q` defined by `Q(v) = f(v,v)/2`. It is defined only in odd characteristic.
+"""
+function corresponding_quadratic_form(B::SesquilinearForm)
+   B.descr==:symmetric || throw(ArgumentError("The form must be a symmetric form"))
+   characteristic(base_ring(B))!=2 || throw(ArgumentError("Corresponding quadratic form not uniquely determined"))
+   M = deepcopy(gram_matrix(B))
+   l = inv(base_ring(B)(2))
+   for i in 1:nrows(M)
+      for j in i+1:nrows(M)
+         M[j,i]=0
+      end
+      M[i,i]*=l
+   end
+   return quadratic_form(M)
+end
+
+
+
+########################################################################
+#
+# Fields of the variable
+#
+########################################################################
+
+"""
+    gram_matrix(B::SesquilinearForm)
+Return the Gram matrix of a sesquilinear or quadratic form `B`.
+"""
+function gram_matrix(f::SesquilinearForm)
+   isdefined(f,:matrix) && return f.matrix
+
+   @assert f.descr==:quadratic && isdefined(f,:pol) "Cannot determine Gram matrix"
+   d = nvars(parent(f.pol))
+   B = zero_matrix( base_ring(f.pol), d, d )
+   V = collect(exponent_vectors(f.pol))
+   C = collect(coeffs(f.pol))
+   for i in 1:length(V)
+      for j in 1:d
+         if V[i][j] !=0
+            global x = j
+            break
+         end
+      end
+      for j in 1:d
+         if V[i][d+1-j] !=0
+            global y = d+1-j
+            break
+         end
+      end
+      B[x,y] = C[i]
+   end
+   f.matrix = B
+   return B
+end
+
+"""
+    defining_polynomial(f::SesquilinearForm)
+Return the polynomial that defines the quadratic form `f`.
+"""
+function defining_polynomial(f::SesquilinearForm)
+   isdefined(f,:pol) && return f.pol
+
+   @assert f.descr == :quadratic "Polynomial defined only for quadratic forms"
+   R = PolynomialRing(base_ring(f.matrix), nrows(f.matrix) )[1]
+   p = zero(R)
+   for i in 1:nrows(f.matrix)
+   for j in i:nrows(f.matrix)
+      p += f.matrix[i,j] * R[i]*R[j]
+   end
+   end
+   f.pol = p
+   return p
+end
+
+
+function assign_from_description(f::SesquilinearForm)
+   if f.descr==:quadratic f.X=GAP.Globals.QuadraticFormByMatrix(f.mat_iso(gram_matrix(f)))
+   elseif f.descr==:symmetric || f.descr==:alternating f.X=GAP.Globals.BilinearFormByMatrix(f.mat_iso(gram_matrix(f)))
+   elseif f.descr==:hermitian f.X=GAP.Globals.HermitianFormByMatrix(f.mat_iso(gram_matrix(f)),f.mat_iso.fr.codomain)
+   else error("unsupported description")
+   end
+end
+
+
+function Base.getproperty(f::SesquilinearForm, sym::Symbol)
+
+   if isdefined(f,sym) return getfield(f,sym) end
+
+   if sym === :mat_iso
+      f.mat_iso = gen_mat_iso(nrows(gram_matrix(f)), base_ring(gram_matrix(f)))
+
+   elseif sym == :X
+      if !isdefined(f, :X)
+         if !isdefined(f,:mat_iso)
+            f.mat_iso = gen_mat_iso(nrows(gram_matrix(f)), base_ring(gram_matrix(f)))
```

Comment:
How often do you think we'll have to create form objects? The complexity here where we create conversions of the matrix lazily really only makes sense when we creates tons of forms quickly. Otherwise, just create the form in both format and be done with it, and perhaps we then don't even need to store the `mat_iso`?

Reply:
I use the GAP underlying object for methods I don't have direct algorithms at the moment (e.g. `witt_index` or `radical`). We can also consider `.X` and `.mat_iso` as `@declare_other` objects.

Reply:
Is the problem the overhead of creating `gen_mat_iso`?
If these objects are cached then this disappears.
(I expect that typically not too many different characteristics and matrix dimensions occur at the same time, and then the number of these objects is small.)

My impression of the "lazy conversions" was that this is a general feature of matrices, and that one need not think about where to use this and where better to avoid it.